cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -145,45 +158,153 @@ func runApply(flags *applyFlags, userVersion string) error {
 		}
 	}
 
-	waiter := getWaiter(flags.dryRun, client)
-
-	// Use a prepuller implementation based on creating DaemonSets
-	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
-	klog.V(1).Infoln("[upgrade/apply] creating prepuller")
-	prepuller := upgrade.NewDaemonSetPrepuller(client, waiter, &cfg.ClusterConfiguration)
-	componentsToPrepull := constants.ControlPlaneComponents
-	if cfg.Etcd.External == nil && flags.etcdUpgrade {
-		componentsToPrepull = append(componentsToPrepull, constants.Etcd)
-	}
-	if err := upgrade.PrepullImagesInParallel(prepuller, flags.imagePullTimeout, componentsToPrepull); err != nil {
-		return errors.Wrap(err, "[upgrade/prepull] Failed prepulled the images for the control plane components error")
-	}
-
 	// Now; perform the upgrade procedure
 	klog.V(1).Infoln("[upgrade/apply] performing upgrade")
-	if err := PerformControlPlaneUpgrade(flags, client, waiter, cfg); err != nil {
+	if err := PerformControlPlaneUpgrade(flags, applyData.Client(), applyData.Waiter(), applyData.Cfg()); err != nil {
 		return errors.Wrap(err, "[upgrade/apply] FATAL")
 	}
 
-	// Upgrade RBAC rules and addons.
-	klog.V(1).Infoln("[upgrade/postupgrade] upgrading RBAC rules and addons")
-	if err := upgrade.PerformPostUpgradeTasks(client, cfg, newK8sVersion, flags.dryRun); err != nil {
+	// Perform postupgrade tasks.
+	klog.V(1).Infoln("[upgrade/postupgrade] running post-upgrade tasks")
+	if err := upgrade.PerformPostUpgradeTasks(applyData.Client(), applyData.Cfg(), applyData.UserVersion(), flags.dryRun); err != nil {
 		return errors.Wrap(err, "[upgrade/postupgrade] FATAL post-upgrade error")
 	}
 
+	// perform addon tasks
+	klog.V(1).Infoln("[upgrade/addons] upgrading addons")
+	if err := addons.UpgradeAddons(applyData.Client(), applyData.Cfg(), flags.dryRun); err != nil {
+		return errors.Wrap(err, "[upgrade/addons] FATAL addons error")
+	}
+
 	if flags.dryRun {
 		fmt.Println("[dryrun] Finished dryrunning successfully!")
 		return nil
 	}
 
 	fmt.Println("")
-	fmt.Printf("[upgrade/successful] SUCCESS! Your cluster was upgraded to %q. Enjoy!\n", cfg.KubernetesVersion)
+	fmt.Printf("[upgrade/successful] SUCCESS! Your cluster was upgraded to %q. Enjoy!\n", applyData.Cfg().KubernetesVersion)
 	fmt.Println("")
 	fmt.Println("[upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets if you haven't already done so.")
 
 	return nil
 }
 
+func newApplyData(flags *applyFlags, userVersion string) (*applyData, error) {
```

Comment:
this builds the struct that is going to be passed to the phases. it took half of what `enforceRequirements` originally did.